bfunwind/src/dwarf4.cpp

```diff
@@ -0,0 +1,575 @@
+        auto value = dwarf4::decode_sleb128(p) * cie.data_alignment();
+        row->set_reg(cfi_register(reg, rule_val_offsetn, value));
+        log("r%ld (%s) at cfa + n(%ld)\n", reg, state->name(reg), value);
+    })
+
+    if_opcode(DW_CFA_val_expression, {
+        ABORT("DWARF4 expressions currently not supported");
+    })
+
+    if_opcode(DW_CFA_GNU_args_size, {
+        ABORT("GNU extension DW_CFA_GNU_args_size is currently not supported");
+    })
+
+    if_opcode(DW_CFA_GNU_negative_offset_extended, {
+        auto reg = dwarf4::decode_uleb128(p);
+        auto value = (int64_t)dwarf4::decode_uleb128(p) * cie.data_alignment();
+        row->set_reg(cfi_register(reg, rule_offsetn, -value));
+        log("r%ld (%s) at cfa + n(%ld)\n", reg, state->name(reg), value);
+    })
+
+    ABORT("unknown cfi opcode");
+}
+
+static void
+private_parse_instructions(cfi_table_row *row,
+                           const ci_entry &cie,
+                           const fd_entry &fde,
+                           register_state *state,
+                           bool is_cie)
+{
+    uint64_t pc_begin = is_cie ? 0 : fde.pc_begin();
+    uint64_t l1 = is_cie ? 0ULL : state->get_ip() - fde.pc_begin();
+    uint64_t l2 = 0ULL;
+
+    char *p = is_cie ? cie.initial_instructions() : fde.instructions();
+    char *end = is_cie ? cie.entry_end() : fde.entry_end();
+
+    while (p < end && l1 >= l2)
+        private_parse_instruction(row, cie, &p, &l1, &l2, pc_begin, state);
+}
+
+cfi_table_row
+private_decode_cfi(const fd_entry &fde, register_state *state)
+{
+    auto row = cfi_table_row();
+    const auto &cie = fde.cie();
+
+    #ifndef DISABLE_LOGGING
+    auto eh_frame = fde.eh_frame();
+    auto cie_offset1 = (uint64_t)cie.entry_start() - (uint64_t)eh_frame.addr;
+    auto cie_size = (uint64_t)(cie.entry_end() - cie.entry_start());
+
+    log("%08lx ", cie_offset1);
+    log("%08lx ", cie_size);
+    log("%08lx ", 0UL);
+    log("CIE\n");
+    #endif
+
+    private_parse_instructions(&row, cie, fde, state, true);
+    log("\n");
+
+    #ifndef DISABLE_LOGGING
+    auto fde_offset1 = (uint64_t)fde.entry_start() - (uint64_t)eh_frame.addr;
+    auto fde_offset2 = (uint64_t)fde.payload_start() - (uint64_t)eh_frame.addr;
+    auto fde_size = (uint64_t)(fde.entry_end() - fde.entry_start());
+    auto pc_begin = fde.pc_begin();
+    auto pc_end = fde.pc_begin() + fde.pc_range();
+
+    log("%08lx ", fde_offset1);
+    log("%08lx ", fde_size);
+    log("%08lx ", fde_offset2);
+    log("FDE ");
+    log("cie=%08lx ", cie_offset1);
+    log("pc=%08lx..%08lx", pc_begin, pc_end);
+    log("\n");
+    #endif
+
+    private_parse_instructions(&row, cie, fde, state, false);
+    log("\n");
+
+    return row;
+}
+
+// -----------------------------------------------------------------------------
+// DWARF 4 Implementation
+// -----------------------------------------------------------------------------
+
+int64_t
+dwarf4::dwarf4::decode_sleb128(char **addr)
```

Comment:
yeah I fixed that in the PR that follows. I have no idea how that compiles, but yeah, it has been fixed.